tools/src/ops_tools/_generate_yaml.py

```diff
@@ -0,0 +1,434 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate charmcraft.yaml config and actions from Python classes."""
+
+from __future__ import annotations
+
+import dataclasses
+import enum
+import logging
+import re
+from typing import (
+    Any,
+    Final,
+    Generator,
+    Mapping,
+    TypedDict,
+    get_args,
+    get_origin,
+    get_type_hints,
+)
+
+from typing_extensions import NotRequired
+
+import ops
+
+from . import _attrdocs
+
+logger = logging.getLogger(__name__)
+
+
+class OptionDict(TypedDict):
+    type: str
+    """The Juju option type."""
+    description: NotRequired[str]
+    default: NotRequired[bool | int | float | str]
+
+
+class ActionDict(TypedDict, total=False):
+    description: str
+    params: dict[str, Any]
+    """A dictionary of parameters for the action."""
+    required: list[str]
+    """A list of required parameters for the action."""
+    additionalProperties: bool
+    """Whether additional properties are allowed in the action parameters."""
+
+
+JUJU_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'int',
+    float: 'float',
+    str: 'string',
+    ops.Secret: 'secret',
+}
+
+# We currently only handle the basic types that we expect to see in real charms.
+# lists and tuples (arrays) are handled without using this mapping.
+JSON_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'integer',
+    float: 'number',
+    str: 'string',
+}
+
+
+def attr_to_default(cls: type[object], name: str) -> Any:
```

Comment:
```suggestion
def attr_to_default(cls: type[object], name: str) -> object:
```
We explicitly check the types after calling this, so we should be fine to use `object` instead of `Any` for better type safety, right?